### ISSUE TITLE
Ephemery Testnet Configs

### ISSUE THREAD

**Issue by atkinsonholly** [ChainSafe/lodestar#6050]
### Problem description

Include configs to easily connect to the ephemeral testnet, "ephemery".

Description of Ephemery, its behaviour and the integration plans by @taxmeifyoucan :
[Ephemery](https://ephemery.dev/) is a new kind of testnet which is intended for short term testing use cases. It automatically resets back to genesis after every given period which enables reclaiming faucet funds, clearing validator set and keeping small state.

Ephemery runs since December 2022 but is mostly run by external tooling. There is ongoing effort to implement specs also within clients, not only externally. All resources for using the testnet are here: https://github.com/ephemery-testnet/ephemery-resources.

Current specs for client requirements can be found in [the draft EIP](https://github.com/taxmeifyoucan/EIPs/blob/d298cdd8eaf47a21e7770e5c6efef870587c924d/EIPS/eip-6916.md).

### Solution description

Add the `--ephemery` config. 
Add additional flags for developer options related to downloading / generating genesis (see below).

### Additional context

**Add Ephemery to the Lodestar network options (cli and config packages)**
- [x] Add ephemery flag

**Additional genesis options**
Default: load genesis from a URL (checkpointz server). Additional [developer] options currently envisaged:

- [x] Use backup download URL (ephemery.dev)
- [x] Use flag to override with custom path to a `genesis.ssz` file provided by user
- [ ] Use flag to generate genesis with script at `packages/ephemery`. Note: user dependency [eth2-testnet-genesis](https://github.com/protolambda/eth2-testnet-genesis). 

**Reset**
- [ ] Research phase ongoing for ephemery reset.

**Comment by atkinsonholly** [comment]
I'll link my draft PRs here shortly

**Comment by nflaig** [comment]
That's a great idea to add first-class support for Ephemery, I quickly looked at this before and I think the ideal solution would be that you only have to set `--network ephemery` and that's it.

My general understanding of how this would work is
- pull genesis data / network config from github ([ephemery-testnet/ephemery-genesis](https://github.com/ephemery-testnet/ephemery-genesis))
- dynamically load network config in Lodestar
- periodically check if there is a new geneses ([check_testnet](https://github.com/ephemery-testnet/ephemery-scripts/blob/9f93fc04ea15d19560445769014fd80043042393/retention.sh#L72-L100))
- restart Lodestar if new genesis is detected ([node-js-process-restart](https://stackoverflow.com/questions/40835187/node-js-process-restart))

I have just quickly looked at this and haven't used Ephemery myself yet. mostly got my ideas from [Ephemery using geth and lodestar](https://github.com/ephemery-testnet/ephemery-scripts/blob/9f93fc04ea15d19560445769014fd80043042393/manual/setup-geth-lodestar.md) and [Retention script](https://github.com/ephemery-testnet/ephemery-scripts/blob/9f93fc04ea15d19560445769014fd80043042393/retention.sh).

One difficulty I see right now is that both the EL and CL need to be coordinated, if only Lodestar supports this it is not gonna be that useful, you would still have to run a separate script in a cronjob to manage the EL. But either way someone has to start pushing this. For initial testing, could use `--execution.engineMock` to not rely on the EL for now.

Would love to hear your thoughts on how this should work in the end from a users point of view

**Comment by atkinsonholly** [comment]
Hi @nflaig that's right, there's other work happening to integrate ephemery into more clients. Geth is next up, and there is also WIP for Lighthouse + Reth. Thanks for the heads up on how to do the initial testing. 
I've now linked my [draft PR](https://github.com/ChainSafe/lodestar/pull/6054) for adding the ephemery network which should address your first bullet - let me know what you think! 
I'm working on a second PR relating to more dynamic genesis options (your second bullet) which will probably mainly be useful for developer users. 
Work on reset is in the early stages (your third + fourth bullets) - draft typescript version of retention script [here](https://github.com/atkinsonholly/lodestar/pull/1/files) (WIP).

**Comment by nflaig** [comment]
> I've now linked my https://github.com/ChainSafe/lodestar/pull/6054 for adding the ephemery network which should address your first bullet - let me know what you think!

Looks good on first glance, will give it more detailed review later. Mostly need to understand the requirements that Ephemery has a bit better to know if this goes in the right direction.

> Work on reset is in the early stages (your third + fourth bullets) - draft typescript version of retention script [here](https://github.com/atkinsonholly/lodestar/pull/1/files) (WIP).

It's a good first PoC but I feel like an ideal solution for this needs to be inherent to Lodestar and not just written in Typescript that in the end just does system calls

This would give us a few advantages
- Lodestar knows where it's data dirs are, no need to have predefined values for this
- Platform agnostic, should work in docker, ideally also windows and not have any dependency on system level tooling like `systemctl`
- Does not rely on any system level permissions (`sudo`)

> restart Lodestar if new genesis is detected ([node-js-process-restart](https://stackoverflow.com/questions/40835187/node-js-process-restart))

I am also still not sure how this would look like, the CLI could act like a process manager by starting the beacon node and validator client in a child process.

For me mostly the question is what the UX of this should look like, if you have a single script which manages all three clients (el + bn + vc) then coordination is easy between clients but a user would have to run another script which possibly requires additional configuration based on client combinations.

On the other hand, if each clients implement an inherent solution it requires more coordination effort, all three clients need to detect a new genesis at around the same time, clean their data dirs and restart. Technically, this should work fine, I think the validator client for example would just restart if there is a config mismatch with then beacon node and at some point, both would be on the same network with new genesis. The implementation effort for this is considerably higher as each client has to implement a native solution but from a user point of view, if we get this right, it would be seamless.

> I'm working on a second PR relating to more dynamic genesis options (your second bullet) which will probably mainly be useful for developer users.

Actually this might already be the case, if the process starts and the database is empty it will pull the genesis state from github and/or checkpoint sync. Another option is to provide a custom genesis file via CLI `--genesisStateFile` flag.

**Comment by philknows** [comment]
> For me mostly the question is what the UX of this should look like, if you have a single script which manages all three clients (el + bn + vc) then coordination is easy between clients but a user would have to run another script which possibly requires additional configuration based on client combinations.

This might a good idea for our https://github.com/ChainSafe/lodestar-quickstart scripts @g11tech since we can hard restart both Lodestar and the selected EL when a new genesis is detected.

**Comment by atkinsonholly** [comment]
Thanks for all the comments, I really appreciate the input!

I've seen the feedback on the PR and I will do a thorough review to make sure I'm being consistent throughout, as well as capturing recent changes to the ephemery specs and adding sufficiently detailed comments.

> Actually this might already be the case, if the process starts and the database is empty it will pull the genesis state from github and/or checkpoint sync. Another option is to provide a custom genesis file via CLI --genesisStateFile flag.

Sounds good. With the addition of a `--genesisStateFile` flag I believe this would address these three checkboxes from my initial description above:
- Add ephemery flag
- Use backup download URL (ephemery.dev)
- Use flag to override with custom path to a genesis.ssz file provided by user

> It's a good first PoC but I feel like an ideal solution for this needs to be inherent to Lodestar and not just written in Typescript that in the end just does system calls

Absolutely, that's understood.  

The single script approach (lodestar-quickstart) could be a useful first step while investigating how clients can integrate their native approach. Would love to ultimately have the benefit of a seamless UX from an inherent solution, but understand this will take more implementation effort as a longer-term goal. It may take a few iterations to get this right.

**Comment by nflaig** [comment]
> I believe this would address these three checkboxes from my initial description above

Yes, I think the initial solution each client has to implement is to make joining the Ephemery as easy as for other networks, just by adding the `--network` flag

> The single script approach (lodestar-quickstart) could be a useful first step while investigating how clients can integrate their native approach. Would love to ultimately have the benefit of a seamless UX from an inherent solution, but understand this will take more implementation effort as a longer-term goal. It may take a few iterations to get this right.

The more I am thinking about this the more I see this solution implemented in tooling that manages the EL / CL, like dappnode, rocketpool, eth-docker etc., as the client acting as a process manager to restart itself might not be a practical for all client implementations.

**Comment by AdedamolaXL** [comment]
Heyy guys, im working on the reset implementation side of this problem. So far, I have done some preliminary work and will like to know if am going about it correctly?

Here is how i have been approaching a native reset within Lodestar: 

- The most important part of the reset is stopping transactions without stopping node operations and then restarting transactions once a new genesis has been downloaded.
- This is done manually with `sudo systemctl`, which is a bad fit as pointed out in this conversation.
- The analogous function responsible for starting and stopping the beacon node is the `beaconHandler`.  
- I think the reset can be carried out using a function similar to `onGracefulShutdown()` from the `beaconHandler`. Let's call it `gracefulSuspension()`.
- While, `onGracefulShutdown()` receives a 'Ctrl + C' signal prior to shutting down node operations. Our function will receive a signal that a new genesis is available and then stop transactions on the beacon node with `db.close()`
- This is what i think`gracefulSuspension()` will look like within the `beaconHandler()` and i will like to know if am approaching it correctly:
```
   gracefulSuspension(async () => {
      abortController.abort();

      if (pruneInvalidSSZObjectsInterval !== null) {
        clearInterval(pruneInvalidSSZObjectsInterval);
      }
    }, logger.info.bind(logger));

    abortController.signal.addEventListener(
      "abort",
      async () => {
        try {
         await db.close();
          // Must explicitly exit process due to potential active handles
          process.exit(1);
        } catch (e) {
          logger.error("Error closing beacon db", {}, e as Error);
        }
      },
      {once: true}
    );
```
**some notes on the function above**: i don't think it's necessary to persist the network identity and enrs so i've removed it. im not sure if clearing the ssz objects is necessary or not (so am keeping it). i have rightfully removed the node.close() because we still need the node active while performing a genesis reset.

Im still thinking about how the `gracefulSuspension()` will look like within the ephemery package:
- It can simply wait for [checkTestnet](https://github.com/atkinsonholly/lodestar/pull/1/files) before executing.
- Or `checkTestnet` can be readapted to be a process manager that sends a SIGTERM signal to `gracefulSuspension()`.

**Comment by AdedamolaXL** [comment]
> The more I am thinking about this the more I see this solution implemented in tooling that manages the EL / CL, like dappnode, rocketpool, eth-docker etc., as the client acting as a process manager to restart itself might not be a practical for all client implementations.

This is intriguing and i haven't quite considered the prospect of external tooling for handling the reset. I will like to look into this, even as i consider the constraints of implementing a native reset within Lodestar.

> This might a good idea for our https://github.com/ChainSafe/lodestar-quickstart scripts @g11tech since we can hard restart both Lodestar and the selected EL when a new genesis is detected.

Will look into this as well.

**Comment by nflaig** [comment]
Thanks for exploring this topic @AdedamolaXL

Regarding your approach described in https://github.com/ChainSafe/lodestar/issues/6050#issuecomment-1775383179. I really don't think there needs to be any modification in the stop procedure that Lodestar implements.

What you really want if there is a genesis reset is to just completely wipe out the current beacon node, that includes
- shutting down the process to drop internal state, peers, chain state etc.
- clean everything inside `dataDir`, which includes the chain-db, peerstore, and anything else the node might persist

After that is done, the node just has to be restarted, and after https://github.com/ChainSafe/lodestar/pull/6054 is implemented the node will automatically load the new genesis data, connect to new peers and be on a completely new chain.

The main requirement is that you need to have another process that manages the Lodestar process and does the restart and cleanup once it detects the new genesis.


> The most important part of the reset is stopping transactions without stopping node operations and then restarting transactions once a new genesis has been downloaded.

Maybe you can elaborate a bit more on this "stopping transactions without stopping node operations"
- what do you mean by transactions
- and what's the difference to node operations

 I haven't looked into detail how a genesis reset on Ephemery works so my answers here might be based on wrong assumptions but since the genesis state changes, the only option I see is to do a hard reset.

**Comment by AdedamolaXL** [comment]
I am a bit clearer on what the requirements for a reset are now. From what you've said, i've deduced that, this is what is needed:

- [ ] periodically check if there is a new genesis ([check_testnet](https://github.com/ephemery-testnet/ephemery-scripts/blob/9f93fc04ea15d19560445769014fd80043042393/retention.sh#L72-L100))
- [ ] shut down the process (if a new genesis is detected) to drop internal state, peers, chain state etc.
- [ ]  clean everything inside dataDir, which includes the chain-db, peerstore, and anything else the node might persist
- [ ] restart Lodestar with new genesis ([node-js-process-restart](https://stackoverflow.com/questions/40835187/node-js-process-restart))

I have a question about this: 
> shutting down the process to drop internal state, peers, chain state etc.

I understand that this is the same as shutting down the beacon node ?
If yes, I'm not too clear on how to go about shutting down the beacon node without modifying the `beaconHandler` except if the user has to send a 'Ctrl + C' signal for a graceful shutdown in this case ?

> The main requirement is that you need to have another process that manages the Lodestar process and does the restart and cleanup once it detects the new genesis.

If I get you correctly the process manager will then incorporate the aforementioned four tasks ?

> Maybe you can elaborate a bit more on this "stopping transactions without stopping node operations"

Forgive me for this, im still a newbie and i tend to mix things up :)
What i meant here was how to keep the node alive while somehow halting validator transactions like staking which seems out of place with this conversation now anyway.

**Comment by nflaig** [comment]
> From what you've said, i've deduced that, this is what is needed:

Yes, those seem to be the major tasks that need to be done

> I understand that this is the same as shutting down the beacon node ?

More or less it's the same

> If yes, I'm not too clear on how to go about shutting down the beacon node without modifying the beaconHandler except if the user has to send a 'Ctrl + C' signal for a graceful shutdown in this case ?

This really depends how it is implemented in the end
1. we could just start the beacon node in a child process as described [here](https://stackoverflow.com/questions/40835187/node-js-process-restart)
2. or we could try do the restart, clean, etc. without restarting the process itself, more details below

Might have to modify the `beaconHandler` in both cases or at least for option 2. It could be possible to handle it even earlier if option 1 is implemented, maybe similar to how we apply the preset
https://github.com/ChainSafe/lodestar/blob/499fd09b0d1cad3ef395c6edaddd1f5e7df6cf55/packages/cli/src/index.ts#L4

> If I get you correctly the process manager will then incorporate the aforementioned four tasks ?

If option 2 is implemented that has to be done in that handler that manages the child process, for option 1 would likely be in `beaconHandler`


> What i meant here was how to keep the node alive while somehow halting validator transactions like staking which seems out of place with this conversation now anyway.

One thing to keep in mind is that you not only have to do this for the beacon node but also the validator client
https://github.com/ChainSafe/lodestar/blob/499fd09b0d1cad3ef395c6edaddd1f5e7df6cf55/packages/cli/src/cmds/validator/handler.ts#L36


> Forgive me for this, im still a newbie and i tend to mix things up :)

No worries, thanks for exploring this topic and happy to provide clarifications where needed

---

### Possible solution without process restart

I was reading up on what is described in the [EIP 6916 reset section](https://github.com/taxmeifyoucan/EIPs/blob/d298cdd8eaf47a21e7770e5c6efef870587c924d/EIPS/eip-6916.md#reset) today. 

Few notes from the spec and how this could be achieved in Lodestar without restarting the whole process.

> Shutdown client services running the network, e.g. p2p communication, beacon service, execution environment

This can be achieved by calling `node.close()`
https://github.com/ChainSafe/lodestar/blob/499fd09b0d1cad3ef395c6edaddd1f5e7df6cf55/packages/cli/src/cmds/beacon/handler.ts#L132

> Client calls a function which discards the current genesis, all chain or beacon data
Clients include feature for purging the db and it might be useful here

Have to close db (which is already done when closing node) and remove all data, most important one is the database (chain-db)
https://github.com/ChainSafe/lodestar/blob/499fd09b0d1cad3ef395c6edaddd1f5e7df6cf55/packages/cli/src/cmds/beacon/paths.ts#L32

>Client triggers the Genesis function (as defined above):
Behaves like a regular client startup when genesis is not present
New genesis is written into db and initialized

This could be achieved by just calling `beaconHandler` with same args and since the data has been cleaned up it will automatically load the new genesis (thanks to https://github.com/ChainSafe/lodestar/pull/6054)
https://github.com/ChainSafe/lodestar/blob/499fd09b0d1cad3ef395c6edaddd1f5e7df6cf55/packages/cli/src/cmds/beacon/handler.ts#L35

> Main network services are started again pointing to the updated genesis

That what happens if we call the beacon handler again which initializes a new beacon node
https://github.com/ChainSafe/lodestar/blob/499fd09b0d1cad3ef395c6edaddd1f5e7df6cf55/packages/cli/src/cmds/beacon/handler.ts#L76


**Potential issues**

The described approach is not really tested. Right now, we only close the beacon node if the process shuts down, restarting the beacon node means restarting the node process. There might be issues due to in-memory caches which are not cleared, or any other pre-calculated values.

**Comment by taxmeifyoucan** [comment]
@nflaig 
> The more I am thinking about this the more I see this solution implemented in tooling that manages the EL / CL, like dappnode, rocketpool, eth-docker etc., as the client acting as a process manager to restart itself might not be a practical for all client implementations.

This would be a correct approach! The EIP defines 2 levels of support in clients - connecting to the network (obtaining genesis upon start) and retaining it (automatic reset). The former is needed for users to interact with the current iteration and the latter is useful mostly for people running infrastructure long term. The assumption is that such providers can solve it using external devops tool, especially if client architecture doesn't allow for a smooth internal restart. The retention script implements this and it becomes even easier with genesis automatically obtained by clients. 

Practically speaking, this means that the genesis and connecting to the network should be the main focus for initial Ephemery support. That's the main goal and PR to be merged. 
Reset is rather a 'nice to have' feature and mostly depends on its feasibility which @AdedamolaXL researches now.

**Comment by AdedamolaXL** [comment]
Thanks for the additions @taxmeifyoucan @nflaig 

I get the precedence of an external tooling option now over an internal reset for clients.

I can start off with this, by making a PR against it:
> This might a good idea for our https://github.com/ChainSafe/lodestar-quickstart scripts @g11tech since we can hard restart both Lodestar and the selected EL when a new genesis is detected.

and then proceed to provide an external tooling option.

**Comment by shrirajpawar4** [comment]
hey @nflaig @philknows 
I'd like to take this on as part of EPF7 prep if thats okay with you. Will open a draft PR once I have the initial structure in place.

**Comment by shrirajpawar4** [comment]
hello @nflaig @philknows @taxmeifyoucan 
I opened a draft PR for the remaining #6050 work after #6054:
#9204

Please let me know if you’d like the scope adjusted.
